transports/quic/src/connection.rs

```diff
@@ -0,0 +1,465 @@
+            }
+        }
+
+        'send_pending: loop {
+            // Sending the pending event to the endpoint. If the endpoint is too busy, we just
+            // stop the processing here.
+            // There is a bit of a question in play here: should we continue to accept events
+            // through `from_endpoint` if `to_endpoint` is busy?
+            // We need to be careful to avoid a potential deadlock if both `from_endpoint` and
+            // `to_endpoint` are full. As such, we continue to transfer data from `from_endpoint`
+            // to the `quinn_proto::Connection` (see above).
+            // However we don't deliver substream-related events to the user as long as
+            // `to_endpoint` is full. This should propagate the back-pressure of `to_endpoint`
+            // being full to the user.
+            if let Some(pending_to_endpoint) = &mut self.pending_to_endpoint {
+                match Future::poll(Pin::new(pending_to_endpoint), cx) {
+                    Poll::Pending => return Poll::Pending,
+                    Poll::Ready(()) => self.pending_to_endpoint = None,
+                }
+            }
+
+            let now = Instant::now();
+
+            // Poll the connection for packets to send on the UDP socket and try to send them on
+            // `to_endpoint`.
+            // FIXME max_datagrams
+            if let Some(transmit) = self.connection.poll_transmit(now, 1) {
+                let endpoint = self.endpoint.clone();
+                debug_assert!(self.pending_to_endpoint.is_none());
+                self.pending_to_endpoint = Some(Box::pin(async move {
+                    // TODO: ECN bits not handled
+                    endpoint
+                        .send_udp_packet(transmit.destination, transmit.contents)
+                        .await;
+                }));
+                continue 'send_pending;
+            }
+
+            // Timeout system.
+            // We break out of the following loop until if `poll_timeout()` returns `None` or if
+            // polling `self.next_timeout` returns `Poll::Pending`.
+            loop {
+                if let Some(next_timeout) = &mut self.next_timeout {
+                    match Future::poll(Pin::new(next_timeout), cx) {
+                        Poll::Ready(when) => {
+                            self.connection.handle_timeout(when);
+                            self.next_timeout = None;
+                        }
+                        Poll::Pending => break,
+                    }
+                } else if let Some(when) = self.connection.poll_timeout() {
+                    if when <= now {
+                        self.connection.handle_timeout(now);
+                    } else {
+                        //let delay = when - now;
+                        self.next_timeout = Some(Timer::at(when));
+                    }
+                } else {
+                    break;
+                }
```

Comment:
```suggestion
              if let Some(next_timeout) = &mut self.next_timeout {
                  match Future::poll(Pin::new(next_timeout), cx) {
                      Poll::Ready(when) => {
                          self.connection.handle_timeout(when);
                          self.next_timeout = None;
                      }
                      Poll::Pending => break,
                  }
              } 
              if let Some(when) = self.connection.poll_timeout() {
                  if when <= now {
                      self.connection.handle_timeout(now);
                  } else {
                      //let delay = when - now;
                      self.next_timeout = Some(Timer::at(when));
                  }
              } 
```

I think we can just write this as two sequential if statements. If I understand it correctly, we just need the loop because after calling `poll_timeout` we might set `next_timeout`. But given that we just checked that this timeout is in the future, I think polling it directly after will always return `Poll::Pending`.